tests/tenderly-simulation-new.ts

```diff
@@ -0,0 +1,534 @@
+          return { slot: candidateSlot, isVyper: true };
+        }
+      }
+    }
+
+    throw new Error(
+      `Could not find a 'balanceOf' mapping slot for token ${token} on chain ${chainId}`,
+    );
+  }
+
+  async findTokenAllowanceSlot(
+    chainId: number,
+    token: string,
+  ): Promise<{ slot: string; isVyper?: boolean }> {
+    const account = ethers.constants.AddressZero;
+    const spender = ethers.constants.AddressZero.slice(0, -2) + '01';
+
+    const allowanceSimulationRequest = this.buildAllowanceSimulationRequest(
+      chainId,
+      token,
+      account,
+      spender,
+    );
+
+    const { id: simulationId } = await this.simulateTransaction(
+      allowanceSimulationRequest,
+    );
+
+    const simulationDetails = await this.getSimulatedTransactionDetails(
+      simulationId,
+    );
+
+    if (!simulationDetails) {
+      throw `No simulation with id ${simulationId} details found`;
+    }
+
+    const callTrace = simulationDetails.transaction.transaction_info.call_trace;
+
+    const sloadCalls = this.getSLOADCalls(callTrace);
+    // token's storage slots that were read during the `allowance` call
+    const readSlots = sloadCalls
+      .map(call => call.storage_slot?.[0])
+      .filter<string>((slot): slot is string => !!slot);
+
+    const startingPoints = [
+      // regular contract
+      0n,
+      // ERC20Upgradeable
+      BigInt(
+        '0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00',
+      ),
+    ];
+
+    for (const init of startingPoints) {
+      for (let i = init; i < init + 1000n; i += 1n) {
+        const candidateSlot = ethers.utils.defaultAbiCoder.encode(
+          ['uint'],
+          [i],
+        );
+        // try solidity
+        const solidityAllowanceSlot =
+          this.calculateSolidityAddressAllowanceSlot(
+            candidateSlot,
+            account,
+            spender,
+          );
+        if (readSlots.includes(solidityAllowanceSlot)) {
+          return { slot: candidateSlot };
+        }
+        // try vyper
+        const vyperAllowanceSlot = this.calculateVyperAddressAllowanceSlot(
+          candidateSlot,
+          account,
+          spender,
+        );
+        if (readSlots.includes(vyperAllowanceSlot)) {
+          return { slot: candidateSlot, isVyper: true };
+        }
+      }
+    }
+
+    throw new Error(
+      `Could not find a 'allowance' mapping slot for token ${token} on chain ${chainId}`,
+    );
+  }
+
+  async getTokenStorageSlots(
+    chainId: number,
+    token: string,
+  ): Promise<TokenStorageSlots> {
+    const normalizedToken = token.toLowerCase();
+
+    if (normalizedToken === ETHER_ADDRESS) {
+      throw new Error('Cannot provide storage slots for native token');
+    }
+
+    const jsonPath = path.join(__dirname, 'token-storage-slots.json');
+
+    const chainSlots = JSON.parse(
+      await fs.readFile(jsonPath, { encoding: 'utf-8' }),
+    ) as Record<number, Record<string, TokenStorageSlots>>;
```

Comment:
We can cache it in local memory (and update this cache each time new token is added to json) to prevent `readFile` each test case

Reply:
added caching in https://github.com/paraswap/paraswap-dex-lib/pull/886/commits/f7780e0cd3b38ce7015ba42b9c4332642d023df3